### ISSUE TITLE
FlatMap between BIOs

### ISSUE THREAD

**Issue by Skotar** [arrow-kt/arrow-fx#149]
```
class OneError
class TwoError

private fun oneEither(): Either<OneError, String> = OneError().left()
private fun twoEither(value: String): Either<TwoError, Int> = value.toInt().right()

fun either() {
   oneEither()
     .flatMap { twoEither(it) }
     .fold(
       { throw Exception(it::class.simpleName) },
       { ... }
     )
}

private fun oneIO(): IO<OneError, String> = IO.raiseError(OneError())
private fun twoIO(value: String): IO<TwoError, Int> = IO.just(value.toInt())

fun io() {
   oneIO()
     .flatMap { twoIO(it) }
     .unsafeRunSyncEither()
     .fold(
       { throw Exception(it::class.simpleName) },
       { ... }
     )
}
```

The function `either` throws `java.lang.Exception: OneError` as I expect. But the function `io` throws `OneError cannot be cast to class TwoError`. When I look at the signature of `flatMap` (`fun <E, A, B, E2 : E> IOOf<E, A>.flatMap(f: (A) -> IOOf<E2, B>): IO<E2, B>`), it seems that it shouldn't even compile (`E2 : E`).

Should BIO behave like Either?

I've been informed to tag @nomisRev and @raulraja.

**Comment by nomisRev** [comment]
This is technically correct, it works with `Monad<IOPartialOf<Any>>`.

And thus uses the extension function:
`fun <E, A, B> Kind<Kind<ForIO, E>, A>.flatMap(arg1: Function1<A, Kind<Kind<ForIO, E>, B>>): IO<E, B>`

(should be imported by `import arrow.fx.extensions.io.monad.flatMap`)

Since `OneError` and `TwoError` share a common upper bound of `Any` the following code is correct.

```kotlin
class OneError
class TwoError

val one = OneError()

fun oneIO(): IO<OneError, String> = IO.raiseError(one)
fun twoIO(value: String): IO<TwoError, Int> = IO.just(value.toInt())

val t: IO<OneError, String> = oneIO()
val s: IO<Any, Int> = t.flatMap { twoIO(it) }

s shouldBe Left(one)
```

Although following fails to compile:

```kotlin
val t: IO<OneError, String> = oneIO()
val s: IO<TwoError, Int> = t.flatMap { twoIO(it) } //<-- Type missmatch
```

**Comment by pakoito** [comment]
It throws with a type mismatch though, even if it finds the upper bound of `Any`, which is what's throwing me off.

**Comment by Skotar** [comment]
Indeed, `arrow.fx.extensions.io.monad.flatMap` works perfectly! Like in monad comprehension. So what is the purpose of `arrow.fx.flatMap`? It seems that `arrow.fx.extensions.io.monad.flatMap` provides the same functionality without `E` upper bound.

From my perspective, it's still interesting why that code compiles and throws an exception at runtime. The signature `fun <E, A, B, E2 : E> IOOf<E, A>.flatMap(f: (A) -> IOOf<E2, B>): IO<E2, B>` indicates that it shouldn't compile - `TwoError` isn't a subtype of `OneError`. But I can be wrong.

I tested the following piece of code:
```
package flatmap.verification

import arrow.fx.IO
import io.kotest.assertions.arrow.either.shouldBeLeft
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import arrow.fx.extensions.io.monad.flatMap as ioMonadFlatMap
import arrow.fx.flatMap as arrowFxFlatMap

class OneError
class TwoError

class AppTest {

    private val one = OneError()

    private fun oneIO(): IO<OneError, String> = IO.raiseError(one)
    private fun twoIO(value: String): IO<TwoError, Int> = IO.just(value.toInt())

    @Test
    fun ioMonadFlatMap() {
        oneIO()
            .ioMonadFlatMap { twoIO(it) }
            .unsafeRunSyncEither() shouldBeLeft one
    }

    @Test
    fun `ioMonadFlatMap and fold`() {
        oneIO()
            .ioMonadFlatMap { twoIO(it) }
            .unsafeRunSyncEither()
            .fold({it}, {it}) shouldBe one
    }

    @Test
    fun arrowFxFlatMap() {
        oneIO()
            .arrowFxFlatMap { twoIO(it) }
            .unsafeRunSyncEither() shouldBeLeft one
    }

    @Test
    fun `arrowFxFlatMap with fold`() {
        oneIO()
            .arrowFxFlatMap { twoIO(it) }
            .unsafeRunSyncEither()
            .fold({it}, {it}) shouldBe one
    }
}
```
`arrowFxFlatMap`, `ioMonadFlatMap` and `ioMonadFlatMap and fold` works as I expect but `arrowFxFlatMap with fold` throws:
```
class flatmap.verification.OneError cannot be cast to class flatmap.verification.TwoError (flatmap.verification.OneError and flatmap.verification.TwoError are in unnamed module of loader 'app')
java.lang.ClassCastException: class flatmap.verification.OneError cannot be cast to class flatmap.verification.TwoError (flatmap.verification.OneError and flatmap.verification.TwoError are in unnamed module of loader 'app')
```
I enclosed the project: [flatmap-verification.zip](https://github.com/arrow-kt/arrow-fx/files/4545235/flatmap-verification.zip).

**Comment by pakoito** [comment]
The one in Monad is correct, the one in master has an unsafe cast

https://github.com/arrow-kt/arrow-fx/blob/master/arrow-fx/src/main/kotlin/arrow/fx/IO.kt#L992

**Comment by aballano** [comment]
Seems this was fixed in 76fe8452c91d5e4aea00bf19d24c4ab94a710681 already, I'm closing the ticket but feel free to reopen if the problem shows again.
